model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java

```diff
@@ -0,0 +1,204 @@
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionTaskWithResult;
+import org.keycloak.models.locking.GlobalLockProvider;
+import org.keycloak.models.locking.LockAcquiringTimeoutException;
+import org.keycloak.models.map.common.DeepCloner;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.models.map.storage.QueryParameters;
+import org.keycloak.models.map.storage.criteria.DefaultModelCriteria;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static org.keycloak.models.map.storage.criteria.DefaultModelCriteria.criteria;
+
+
+/**
+ * Implementing a {@link GlobalLockProvider} based on a map storage.
+ * This requires the map store to support the entity type {@link MapLockEntity}. One of the stores which supports
+ * this is the JPA Map Store. The store needs to support the uniqueness of entries in the lock area, see
+ * {@link #lock(String)} for details.
+ *
+ * @author Alexander Schwartz
+ */
+public class MapGlobalLockProvider implements GlobalLockProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapGlobalLockProvider.class);
+
+    private final KeycloakSession session;
+    private final long defaultTimeoutMilliseconds;
+    private MapKeycloakTransaction<MapLockEntity, MapLockEntity> tx;
+
+    /**
+     * The lockStoreSupplier allows the store to be initialized lazily and only when needed: As this provider is initialized
+     * for both the outer and the inner transactions, and the store is needed only for the inner transactions.
+     */
+    private final Supplier<MapStorage<MapLockEntity, MapLockEntity>> lockStoreSupplier;
+
+    public MapGlobalLockProvider(KeycloakSession session, long defaultTimeoutMilliseconds, Supplier<MapStorage<MapLockEntity, MapLockEntity>> lockStoreSupplier) {
+        this.defaultTimeoutMilliseconds = defaultTimeoutMilliseconds;
+        this.session = session;
+        this.lockStoreSupplier = lockStoreSupplier;
+    }
+
+    @Override
+    public <V> V withLock(String lockName, Duration timeToWaitForLock, KeycloakSessionTaskWithResult<V> task) throws LockAcquiringTimeoutException {
+        MapLockEntity lockEntity = null;
+        try {
+            if (timeToWaitForLock == null) {
+                // Set default timeout if null provided
+                timeToWaitForLock = Duration.ofMillis(defaultTimeoutMilliseconds);
+            }
+            long started = System.currentTimeMillis();
+            String keycloakInstanceIdentifier = null;
+            Instant timeWhenAcquired = null;
+            while (true) {
```

Comment:
Should we use `Retry.executeWithBackoff`? Similarly as [here](https://github.com/keycloak/keycloak/blob/main/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/locking/HotRodLocksUtils.java#L69)?

Reply:
Thanks for this hint, I've update the code in the most recent commit. I'm now waiting for the build to complete.